rclcpp/test/rclcpp/test_typesupport_helpers.cpp

```diff
@@ -0,0 +1,104 @@
+// Copyright 2018, Bosch Software Innovations GmbH.
+// Copyright 2021, Apex.AI Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <gmock/gmock.h>
+
+#include <memory>
+#include <string>
+#include <utility>
+
+#include "rcpputils/shared_library.hpp"
+
+#include "rclcpp/typesupport_helpers.hpp"
+
+using namespace ::testing;  // NOLINT
+
+TEST(TypesupportHelpersTest, throws_exception_if_filetype_has_no_type) {
+  EXPECT_ANY_THROW(rclcpp::extract_type_identifier("just_a_package_name"));
+}
+
+TEST(TypesupportHelpersTest, throws_exception_if_filetype_has_slash_at_the_start_only) {
+  EXPECT_ANY_THROW(rclcpp::extract_type_identifier("/name_with_slash_at_start"));
+}
+
+TEST(TypesupportHelpersTest, throws_exception_if_filetype_has_slash_at_the_end_only) {
+  EXPECT_ANY_THROW(rclcpp::extract_type_identifier("name_with_slash_at_end/"));
+}
+
+TEST(TypesupportHelpersTest, separates_into_package_and_name_for_correct_package) {
+  std::string package;
+  std::string middle_module;
+  std::string name;
+  std::tie(package, middle_module, name) = rclcpp::extract_type_identifier("package/name");
+
+  EXPECT_THAT(package, StrEq("package"));
+  EXPECT_THAT(middle_module, StrEq(""));
+  EXPECT_THAT(name, StrEq("name"));
+}
+
+TEST(TypesupportHelpersTest, separates_into_package_and_name_for_multiple_slashes) {
+  std::string package;
+  std::string middle_module;
+  std::string name;
+  std::tie(package, middle_module, name) =
+    rclcpp::extract_type_identifier("package/middle_module/name");
+
+  EXPECT_THAT(package, StrEq("package"));
+  EXPECT_THAT(middle_module, StrEq("middle_module"));
+  EXPECT_THAT(name, StrEq("name"));
+}
+
+TEST(TypesupportHelpersTest, throws_exception_if_library_is_not_initialized) {
+  std::shared_ptr<rcpputils::SharedLibrary> library;
+  EXPECT_THROW(
+    rclcpp::get_typesupport_library("invalid/message", "rosidl_typesupport_cpp"),
+    std::runtime_error);
+}
+
+TEST(TypesupportHelpersTest, throws_exception_if_library_cannot_be_found) {
+  EXPECT_THROW(
+    rclcpp::get_typesupport_library("invalid/message", "rosidl_typesupport_cpp"),
+    std::runtime_error);
+}
```

Comment:
I don't understand what the difference is between these two tests. They look the same to me; I guess one can be removed.

Reply:
Remove one of them in be5453c

Reply:
Maybe a copy-paste mistake. The names imply a related, but different error case. Maybe the intention was to change one of them. But yes, right now I agree these are identical and one could be removed.